File: packages/algorithm/leecode100/LRU.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

struct Node {
    int key, value;
    Node* prev;
    Node* next;
    Node(int k, int v): key(k), value(v), prev(nullptr), next(nullptr) {}
};

class LRUCache {
private:
    unordered_map<int, Node*> hash;
    Node* head;
    Node* tail;
    int capacity;
    int size;

public:
    LRUCache(int capacity): capacity(capacity), size(0) {
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head->next = tail;
        tail->prev = head;
    }

    int get(int key) {
        if (hash.find(key) == hash.end()) return -1;
        
        Node* node = hash[key];
        moveToHead(node);
        
        return node->value;
    }

    void put(int key, int value) {
        if (hash.find(key) != hash.end()) {
            Node* node = hash[key];
            node->value = value;
            moveToHead(node);
        } else {
            Node* newNode = new Node(key, value);
            hash[key] = newNode;
            addToHead(newNode);
            size++;
            
            if (size > capacity) {
                Node* node = removeTail();
                hash.erase(node->key);
                delete node;
                size--;
            }
        }
    }

private:
    void moveToHead(Node* node) {
        removeNode(node);
        addToHead(node);
    }

    void removeNode(Node* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }

    void addToHead(Node* node) {
        node->prev = head;
        node->next = head->next;
        
        head->next->prev = node;
        head->next = node;
    }

    Node* removeTail() {
        Node* node = tail->prev;
        removeNode(node);
        return node;
    }
};

int main() {
    LRUCache cache(2);
    cache.put(1, 1);
    cache.put(2, 2);
    cout << cache.get(1) << endl;  // 输出 1
    cache.put(3, 3);  // 逐出 key 为 2 的缓存
    cout << cache.get(2) << endl;  // 输出 -1
    cache.put(4, 4);  // 逐出 key 为 1 的缓存
    cout << cache.get(1) << endl;  // 输出 -1
    cout << cache.get(3) << endl;  // 输出 3
    cout << cache.get(4) << endl;  // 输出 4
    return 0;
}

File: Design/data-structure/chache/LRU.ts
class DoubleLinkedList {
    val: number;
    key: number;
    prev: DoubleLinkedList | null;
    next: DoubleLinkedList | null;
    constructor(key: number ,val: number) {
        this.val = val;
        this.key = key;
        this.prev = null;
        this.next = null;
    }

}

class LRUCache {
    map: Map<number, DoubleLinkedList>;
    head: DoubleLinkedList;
    tail: DoubleLinkedList;
    capacity: number;
    size: number;
    constructor(capacity: number){
        this.map = new Map();
        this.capacity = capacity;
        this.size = 0;
        this.head = new DoubleLinkedList(-1, -1);
        this.tail = new DoubleLinkedList(-1, -1);

        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    get(key: number): number {
        if (!this.map.has(key)) {
            return -1;
        }

        const node = this.map.get(key)!;
        this.removeNode(node);
        this.addToHead(node);

        return node.val;
    }

    put(key: number, value: number): void {
        if (this.map.has(key)) {
            const node = this.map.get(key)!;
            node.val = value;
            this.removeNode(node);
            this.addToHead(node);
        } else {
            const newNode = new DoubleLinkedList(key, value);
            this.addToHead(newNode);
            this.map.set(key, newNode);
            this.size ++;

            if (this.size > this.capacity) {
                const lastNode = this.tail.prev!;

                this.removeNode(lastNode);
                this.map.delete(lastNode.key);
            }
        }
    }

    // 双向链表相关方法
    addToHead(node: DoubleLinkedList) {
        this.head.next!.prev = node;
        node.next = this.head.next;
        node.prev = this.head;
        this.head.next = node;
    }

    removeNode(node: DoubleLinkedList) {
        node.prev!.next = node.next;
        node.next!.prev = node.prev;
    }

}
